# Speech and Language Processing (SLP)
## Master Systèmes Embarqués et Traitement de l'Information (SETI)


---------------------------

### Project: **Sarcasm Detection in Natural Language Processing Applied to English Text**
### Alaf DO NASCIMENTO SANTOS
### Ianis GIRAUD
---------------------------

## Introduction

As a final project for the Speech and Language Processing course, and first actual AI project for the team, we decided to apply the acquired knowledge to a simple Natural Language Processing (NLP) task, principally in the area of text mining. We aim to train a model capable of recognising sarcasm in English texts.

In the complex landscape of NLP, our AI project, titled "Sarcasm Detection in Natural Language Processing Applied to English Text" addresses the challenge of deciphering sarcasm within individual sentences. Our goal is to develop a model capable of providing a binary output—indicating whether a given sentence is sarcastic (true) or not (false).

Unlike more powerful systems, our focus is on simplicity and resource efficiency, especially important when dealing with text mining tasks. Sarcasm, a form of verbal irony, adds an extra layer of complexity to the already challenging field of text analysis. In this context, the challenge extends to navigating the nuanced aspects of English expressions.

Our approach integrates fundamental linguistic analysis with machine learning techniques. The model will be trained on datasets that capture instances of sarcasm within English sentences, considering the specific linguistic nuances and cultural context from the social media Reddit through the Self-Annotated Reddit Corpus (SARC) [1]. This training process aims to equip our AI system with the ability to discern sarcasm efficiently, even when presented with the constraints of text mining tasks.

The potential applications of our system are numerous, ranging from improving sentiment analysis in customer feedback to enhancing the understanding of social media interactions. By highlighting the sentence-level analysis, our project strikes a balance between efficiency and effectiveness, offering a pragmatic solution for scenarios where processing entire texts in social media may be challenging.

## Installing and Importing necessary packages

In [33]:
!pip3 install matplotlib
!pip3 install portalocker==2.8.2
!pip3 install pandas
%matplotlib inline

Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable


In [34]:
import pandas as pd
import torch
import time
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from tqdm.notebook import tqdm
from torchtext.data.utils import get_tokenizer
from torchtext.vocab import build_vocab_from_iterator
from torch.utils.data import DataLoader
from torch import nn
from torch.utils.data.dataset import random_split
from torchtext.data.functional import to_map_style_dataset

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

cuda


## Preprocessing

The SARC dataset can be found at <https://nlp.cs.princeton.edu/old/SARC/2.0/>. It is a very large dataset, where each statement is self-annotated (sarcasm is labeled by the author) and provides the user, topic, and conversation context. In order to reduce the size of our dataset and only work with the target information, while keeping a limited complexity, we performed a preprocessing step in C code. The following C code was used to clean the json file containing the corpus raw data.


```
#include <stdio.h>


int main(void) {
    int c;
    int n;
    int level = 0;

    while ((c = getchar()) != EOF) {
        if (c=='[' || c=='{') {
            putchar(c);
            putchar('\n');
            level++;
            for (int i=0; i<level; i++) {
                putchar('\t');
            }
        } else if (c==']' || c=='}') {
            putchar('\n');
            level--;
            for (int i=0; i<level; i++) {
                putchar('\t');
            }
            putchar(c);
        } else if (c==',') {
            putchar(c);
            putchar('\n');
            for (int i=0; i<level; i++) {
                putchar('\t');
            }
        } else {
            putchar(c);
        }
    }
    return 0;
}
```

EXPLAIN HERE HOW WE PASSED FROM ONE UNIQUE JSON TO TWO CSVs FILES (I DIDN'T GET)


Since we are not supposed to include external data directly in our submission, the clean corpus dataset has been uploaded to a personal GitHub repository. In the Python code, we access the files link hosted on Github and we load it through the **read_csv** function.

In [35]:
url_train = "https://raw.githubusercontent.com/alafSantos/SETI-SLP-Sarcasm-recogniser/main/SARC/merged-train.csv"
url_train = "SARC/merged-train.csv"
df_train = pd.read_csv(url_train, usecols=["label", "text_parent", "text_post"], encoding='utf-8')
train_iter = df_train.iterrows()
df_train.head()


label                                        text_parent  \
0      1  I've been searching for the answer for this fo...   
1      0  I've been searching for the answer for this fo...   
2      1  Michael Phelps Apologizes For "Regrettable" Be...   
3      0  Michael Phelps Apologizes For "Regrettable" Be...   
4      0  Utah wants to create a database to track the i...   

                                           text_post  
0                      Religion must have the answer  
1  It's obviously tracks from a giant water tract...  
2  Wow...he smoked pot...oh lord hes such a horri...  
3  Wow, his girlfriend is uhm... Ah fuck it, he's...  
4  I think the government should track every morm...

In [36]:
url_test = "https://raw.githubusercontent.com/alafSantos/SETI-SLP-Sarcasm-recogniser/main/SARC/merged-test.csv"
url_test = "SARC/merged-test.csv"
df_test = pd.read_csv(url_test, usecols=["label", "text_parent", "text_post"], encoding='utf-8')
test_iter = df_test.iterrows()
df_test.head()

label                                        text_parent  \
0      1  The vast majority of Republicans rallied behin...   
1      0  The vast majority of Republicans rallied behin...   
2      1  "...two-income families often have even less i...   
3      0  "...two-income families often have even less i...   
4      1                           Heath Ledger Wins Oscar!   

                                           text_post  
0        Yes, cuz tax cuts will help those w/o jobs!  
1        If cutting taxes fails... cut taxes harder.  
2  Chalk it up to the ever-increasing cost of fre...  
3  We're about to finally get affordable housing,...  
4   oh wow I am so surprised I never saw this coming

EXPLIQUER

In [37]:
tokenizer = get_tokenizer("basic_english")

def yield_tokens(data_iter):
    for _, row in data_iter:
        text = str(row["text_post"])
        yield tokenizer(text)

vocab = build_vocab_from_iterator(yield_tokens(train_iter), specials=["<unk>"])
vocab.set_default_index(vocab["<unk>"])

text_pipeline = lambda x: vocab(tokenizer(x))
label_pipeline = lambda x: int(x)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

## NO CONTEXT APPROACH

In [38]:
def collate_batch(batch):
    label_list, text_list, offsets = [], [], [0]
    for index, row in batch:
        _label = row["label"]
        _text = str(row["text_post"])

        label_list.append(label_pipeline(_label))
        processed_text = torch.tensor(text_pipeline(_text), dtype=torch.int64)
        text_list.append(processed_text)
        offsets.append(processed_text.size(0))
    label_list = torch.tensor(label_list, dtype=torch.int64)
    offsets = torch.tensor(offsets[:-1]).cumsum(dim=0)
    text_list = torch.cat(text_list)
    return label_list.to(device), text_list.to(device), offsets.to(device)


dataloader = DataLoader(
    train_iter, batch_size=8, shuffle=False, collate_fn=collate_batch
)

In [39]:
class TextClassificationModel(nn.Module):
    def __init__(self, vocab_size, embed_dim, num_class):
        super(TextClassificationModel, self).__init__()
        self.embedding = nn.EmbeddingBag(vocab_size, embed_dim, sparse=False)
        self.fc = nn.Linear(embed_dim, num_class)
        self.init_weights()

    def init_weights(self):
        initrange = 0.5
        self.embedding.weight.data.uniform_(-initrange, initrange)
        self.fc.weight.data.uniform_(-initrange, initrange)
        self.fc.bias.data.zero_()

    def forward(self, text, offsets):
        embedded = self.embedding(text, offsets)
        return self.fc(embedded)


In [40]:
num_class = 2 #1 or 0 always
vocab_size = len(vocab)
emsize = 64
model = TextClassificationModel(vocab_size, emsize, num_class).to(device)

## Training and Evaluation

In [41]:
def train(dataloader, model=model):
    model.train()
    total_acc, total_count = 0, 0
    log_interval = 500
    start_time = time.time()

    for idx, (label, text, offsets) in enumerate(dataloader):
        optimizer.zero_grad()
        predicted_label = model(text, offsets)
        loss = criterion(predicted_label, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.1)
        optimizer.step()
        total_acc += (predicted_label.argmax(1) == label).sum().item()
        total_count += label.size(0)
        if idx % log_interval == 0 and idx > 0:
            elapsed = time.time() - start_time
            print(
                "| epoch {:3d} | {:5d}/{:5d} batches "
                "| accuracy {:8.3f}".format(
                    epoch, idx, len(dataloader), total_acc / total_count
                )
            )
            total_acc, total_count = 0, 0
            start_time = time.time()

def evaluate(dataloader, model=model):
    model.eval()
    total_acc, total_count = 0, 0
    with torch.no_grad():
        for _, (label, text, offsets) in enumerate(dataloader):
            predicted_label = model(text, offsets)
            total_acc += (predicted_label.argmax(1) == label).sum().item()
            total_count += label.size(0)
    return total_acc / total_count


            
# Hyperparameters
EPOCHS = 10  # epoch
LR = 5  # learning rate
BATCH_SIZE = 64  # batch size for training

criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=LR)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1.0, gamma=0.1)
total_accu = None

train_iter = df_train.iterrows()
test_iter = df_test.iterrows()

train_dataset = to_map_style_dataset(train_iter)
test_dataset = to_map_style_dataset(test_iter)
num_train = int(len(train_dataset) * 0.95)
split_train_, split_valid_ = random_split(
    train_dataset, [num_train, len(train_dataset) - num_train]
)

train_dataloader = DataLoader(
    split_train_, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_batch
)
valid_dataloader = DataLoader(
    split_valid_, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_batch
)
test_dataloader = DataLoader(
    test_dataset, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_batch
)

for epoch in range(0, EPOCHS):
    epoch_start_time = time.time()
    train(train_dataloader)
    accu_val = evaluate(valid_dataloader)
    if total_accu is not None and total_accu > accu_val:
        scheduler.step()
    else:
        total_accu = accu_val
    print("-" * 59)
    print(
        "| end of epoch {:3d} | time: {:5.2f}s | "
        "valid accuracy {:8.3f} ".format(
            epoch, time.time() - epoch_start_time, accu_val
        )
    )
    print("-" * 59)

| epoch   0 |   500/ 3817 batches | accuracy    0.565
| epoch   0 |  1000/ 3817 batches | accuracy    0.595
| epoch   0 |  1500/ 3817 batches | accuracy    0.608
| epoch   0 |  2000/ 3817 batches | accuracy    0.607
| epoch   0 |  2500/ 3817 batches | accuracy    0.613
| epoch   0 |  3000/ 3817 batches | accuracy    0.622
| epoch   0 |  3500/ 3817 batches | accuracy    0.615
-----------------------------------------------------------
| end of epoch   0 | time: 12.29s | valid accuracy    0.628 
-----------------------------------------------------------
| epoch   1 |   500/ 3817 batches | accuracy    0.633
| epoch   1 |  1000/ 3817 batches | accuracy    0.632
| epoch   1 |  1500/ 3817 batches | accuracy    0.633
| epoch   1 |  2000/ 3817 batches | accuracy    0.637
| epoch   1 |  2500/ 3817 batches | accuracy    0.631
| epoch   1 |  3000/ 3817 batches | accuracy    0.637
| epoch   1 |  3500/ 3817 batches | accuracy    0.631
-----------------------------------------------------------
| e

Checking the results of test dataset

In [42]:
print("Checking the results of test dataset.")
accu_test = evaluate(test_dataloader)
print("test accuracy {:8.2f}".format(accu_test*100)+"%")

Checking the results of test dataset.


test accuracy    65.27%


In [43]:
def predict_no_context(text, text_pipeline=text_pipeline):
    with torch.no_grad():
        text = torch.tensor(text_pipeline(text))
        output = model(text, torch.tensor([0]))
        return output.argmax(1).item() == 1

## Context Approach

In [44]:
text_pipeline = lambda x, y: (vocab(tokenizer(str(x))), vocab(tokenizer(str(y))))
label_pipeline = lambda x: int(x)

def collate_batch(batch):
    label_list, text_parent_list, text_post_list = [], [], []
    offsets_parent, offsets_post = [0], [0]
    for index, row in batch:
        label = row['label']
        text_post = row['text_post']
        text_parent = row['text_parent']
        label_list.append(label_pipeline(label))
        tokens_parent, tokens_post = text_pipeline(text_parent, text_post)
        processed_text_parent = torch.tensor(tokens_parent, dtype=torch.int64)
        processed_text_post = torch.tensor(tokens_post, dtype=torch.int64)
        text_parent_list.append(processed_text_parent)
        text_post_list.append(processed_text_post)
        offsets_parent.append(processed_text_parent.size(0))
        offsets_post.append(processed_text_post.size(0))
    label_list = torch.tensor(label_list, dtype=torch.float)
    offsets_parent = torch.tensor(offsets_parent[:-1]).cumsum(dim=0)
    offsets_post = torch.tensor(offsets_post[:-1]).cumsum(dim=0)
    text_parent_list = torch.cat(text_parent_list)
    text_post_list = torch.cat(text_post_list)
    return (label_list.to(device), text_parent_list.to(device), text_post_list.to(device), 
            offsets_parent.to(device), offsets_post.to(device))

train_iter = df_train.iterrows()
dataloader = DataLoader(
    train_iter, batch_size=8, shuffle=False, collate_fn=collate_batch
)

In [45]:
class TextClassificationModel(nn.Module):
    def __init__(self, vocab_size, embed_dim, embed_dim_post, hidden_dim):
        super(TextClassificationModel, self).__init__()
        self.embedding_parent = nn.EmbeddingBag(vocab_size, embed_dim, sparse=False)
        self.embedding_post = nn.EmbeddingBag(vocab_size, embed_dim_post, sparse=False)
        self.hidden = nn.Linear(embed_dim*2, hidden_dim)
        self.fc = nn.Linear(hidden_dim, 1)
        self.init_weights()

    def init_weights(self):
        initrange = 0.5
        self.embedding_parent.weight.data.uniform_(-initrange, initrange)
        self.embedding_post.weight.data.uniform_(-initrange, initrange)
        self.hidden.weight.data.uniform_(-initrange, initrange)
        self.hidden.bias.data.zero_()
        self.fc.weight.data.uniform_(-initrange, initrange)
        self.fc.bias.data.zero_()

    def forward(self, text_parent, text_post, offsets_parent, offsets_post):
        embedded_parent = self.embedding_parent(text_parent, offsets_parent)
        embedded_post = self.embedding_parent(text_post, offsets_post)
        embedded = torch.hstack([embedded_parent,embedded_post])
        middle = self.hidden(torch.sigmoid(embedded))
        output = self.fc(torch.sigmoid(middle))
        return torch.sigmoid(output).squeeze()

In [46]:
train_iter = df_test.iterrows()
emsize = 64
embed_dim_post = 96
hidden_size = 64
model2 = TextClassificationModel(vocab_size, emsize, embed_dim_post, hidden_size).to(device)

### Training and Evaluation

In [47]:
def train(dataloader, model=model2):
    model.train()
    total_acc, total_count = 0, 0
    log_interval = 500
    start_time = time.time()

    for idx, (label, text_parent, text_post, offsets_parent, offsets_post) in enumerate(dataloader):
        optimizer.zero_grad()
        predicted_label = model(text_parent, text_post, offsets_parent, offsets_post)
        loss = criterion(predicted_label, label)
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.1)
        optimizer.step()
        predicted_label[predicted_label > 0.5] = 1
        predicted_label[predicted_label <= 0.5] = 0
        total_acc += (predicted_label == label).sum().item()
        total_count += label.size(0)
        if idx % log_interval == 0 and idx > 0:
            elapsed = time.time() - start_time
            print(
                "| epoch {:3d} | {:5d}/{:5d} batches "
                "| accuracy {:8.3f}".format(
                    epoch, idx, len(dataloader), total_acc / total_count
                )
            )
            total_acc, total_count = 0, 0
            start_time = time.time()


def evaluate(dataloader, model=model2):
    model.eval()
    total_acc, total_count = 0, 0

    with torch.no_grad():
        for idx, (label, text_parent, text_post, offsets_parent, offsets_post) in enumerate(dataloader):
            predicted_label = model(text_parent, text_post, offsets_parent, offsets_post)
            loss = criterion(predicted_label, label)
            predicted_label[predicted_label > 0.5] = 1
            predicted_label[predicted_label <= 0.5] = 0
            total_acc += (predicted_label == label).sum().item()
            total_count += label.size(0)
    return total_acc / total_count

# Hyperparameters
EPOCHS = 10  # epoch
LR = 5  # learning rate
BATCH_SIZE = 64  # batch size for training

criterion = torch.nn.MSELoss()
optimizer = torch.optim.SGD(model2.parameters(), lr=LR)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1.0, gamma=0.1)
total_accu = None
train_iter = df_train.iterrows()
test_iter = df_test.iterrows()
train_dataset = to_map_style_dataset(train_iter)
test_dataset = to_map_style_dataset(test_iter)
num_train = int(len(train_dataset) * 0.95)
split_train_, split_valid_ = random_split(
    train_dataset, [num_train, len(train_dataset) - num_train]
)

train_dataloader = DataLoader(
    split_train_, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_batch
)
valid_dataloader = DataLoader(
    split_valid_, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_batch
)
test_dataloader = DataLoader(
    test_dataset, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_batch
)

print("Starting training!")
for epoch in range(0, EPOCHS):
    epoch_start_time = time.time()
    train(train_dataloader)
    accu_val = evaluate(valid_dataloader)
    if total_accu is not None and total_accu > accu_val:
        scheduler.step()
    else:
        total_accu = accu_val
    print("-" * 59)
    print(
        "| end of epoch {:3d} | time: {:5.2f}s | "
        "valid accuracy {:8.3f} ".format(
            epoch, time.time() - epoch_start_time, accu_val
        )
    )
    print("-" * 59)

Starting training!
| epoch   0 |   500/ 3817 batches | accuracy    0.494
| epoch   0 |  1000/ 3817 batches | accuracy    0.503
| epoch   0 |  1500/ 3817 batches | accuracy    0.499
| epoch   0 |  2000/ 3817 batches | accuracy    0.500
| epoch   0 |  2500/ 3817 batches | accuracy    0.505
| epoch   0 |  3000/ 3817 batches | accuracy    0.498
| epoch   0 |  3500/ 3817 batches | accuracy    0.501
-----------------------------------------------------------
| end of epoch   0 | time: 17.95s | valid accuracy    0.500 
-----------------------------------------------------------
| epoch   1 |   500/ 3817 batches | accuracy    0.506
| epoch   1 |  1000/ 3817 batches | accuracy    0.513
| epoch   1 |  1500/ 3817 batches | accuracy    0.518
| epoch   1 |  2000/ 3817 batches | accuracy    0.517
| epoch   1 |  2500/ 3817 batches | accuracy    0.521
| epoch   1 |  3000/ 3817 batches | accuracy    0.530
| epoch   1 |  3500/ 3817 batches | accuracy    0.540
--------------------------------------------

In [49]:
print("Checking the results of test dataset.")
accu_test = evaluate(test_dataloader)
print("test accuracy {:8.2f}".format(accu_test*100)+"%")

Checking the results of test dataset.
test accuracy    60.69%


In [50]:
def predict_context(text_parent, text_post, text_pipeline=text_pipeline):
    with torch.no_grad():
        text_parent, text_post = text_pipeline(text_parent, text_post)
        output = model2( torch.tensor(text_parent), torch.tensor(text_post), 
                        torch.tensor([0]), torch.tensor([0]) )
        return output.item() > 0.5

## VADER Sentiment Analysis Approach

With VADER approach we can take all the words in a sentence and then set a value, either negative, positive or neutral and combines those values in order to tell us if the sentence itself is more globally positive or negative. VADER gives us values from -1 to 1 for each word and a global compound (which we will be using here). From [6] we have a study showing that most of the time sarcasm comes with positive sentences, but in a negative situation or context. For simplicity, we will be using VADER as a voter in our final decision by defining a sentence as sarcastic when it has a positive or neutral compound in the answer, otherwise it will be non sarcastic; together with a negative or neutral context. It is important to highlight that VADER doesn't take into account the relationship between the words, which is very important in the real world. But it will be used here as voter for simplicity.

In [51]:
test_iter = df_test.iterrows()

def evaluate():
    total_acc, total_count = 0, 0

    for _, row in test_iter:
        label = row['label']
        text_post = str(row['text_post'])
        text_parent = str(row['text_parent'])
        predicted_label = predict_vader(text_parent, text_post)
        total_acc += (predicted_label == (label==1))
        total_count += 1
    
    return total_acc / total_count


def predict_vader(context, answer):
    sia = SentimentIntensityAnalyzer()
    vader1_result = sia.polarity_scores(context)["compound"]
    vader2_result = sia.polarity_scores(answer)["compound"]
    result = (vader1_result <= 0 and vader2_result >= 0)
    return result


print("Checking the results of test dataset.")
accu_test = evaluate()
print("test accuracy {:8.2f}".format(accu_test*100)+"%")

Checking the results of test dataset.
test accuracy    50.38%


## Demonstration

In the end, to demonstrate our sarcasm recognition system, we have a voting system. Each approach developed will give a vote, 0 for non sarcastic, 1 for sarcastic, and if the sum of the votes is greater than 1 (2 or 3) it means that the input is sarcastic.

In [52]:
# https://www.yourdictionary.com/articles/examples-sarcasm-meaning-types
input1 = [
    "When something bad happens",
    "When you expected something to happen, especially after warning someone about it",
    "Mine operator in accident had 57 safety violations last month alone and and did not install a required ventilation system that would have prevented the disaster"
    "When a roommate is acting bizarre",
    "When someone says something that is very obvious"
          ]


input2 = [
    "That's just what I needed today!",
    "Well, what a surprise.",
    "How does this compare to other mine operations"
    "Is it time for your medication or mine?",
    "Really, Sherlock? No! You are clever."
          ]

model = model.to("cpu")
model2 = model2.to("cpu")

for i in range(0, len(input1)):
    noContext_vote = predict_no_context(input2[i])
    context_vote = predict_context(input1[i], input2[i])
    vader_vote = predict_vader(input1[i], input2[i])
    result = sum([noContext_vote, context_vote, vader_vote]) > 1

    print("--------------------\nSituation ", i, ":", input1[i])
    print("Remark: ", input2[i])
    print("M1: ", noContext_vote)
    print("M2: ", context_vote)
    print("M3: ", vader_vote)
    print("Voting Result: ", result)

--------------------
Situation  0 : When something bad happens
Remark:  That's just what I needed today!
M1:  False
M2:  True
M3:  True
Voting Result:  True
--------------------
Situation  1 : When you expected something to happen, especially after warning someone about it
Remark:  Well, what a surprise.
M1:  True
M2:  True
M3:  True
Voting Result:  True
--------------------
Situation  2 : Mine operator in accident had 57 safety violations last month alone and and did not install a required ventilation system that would have prevented the disasterWhen a roommate is acting bizarre
Remark:  How does this compare to other mine operationsIs it time for your medication or mine?
M1:  False
M2:  False
M3:  True
Voting Result:  False
--------------------
Situation  3 : When someone says something that is very obvious
Remark:  Really, Sherlock? No! You are clever.
M1:  True
M2:  True
M3:  True
Voting Result:  True


## Conclusion

    Throughout our work, we developed 3 approaches for sarcasm recognition so we were able to apply the Condorcet's jury theorem, so we obtained a more trustful and robust system to give us binary predictions in either a setence is sarcastic or not. We have worked with the SARC corpus, containing Reddit comments self annotated by the authors.

    For the first model, we designed a simple sarcasm recoginition system, taking only a setence as input for the analysis (no context), based on a Pytorch example for sentiment analysis. There we had a test accuracy over 65%.

    Concerning the second model, using the same self-annotated dataset, we made some modifications in the first model in order to take into account the context in which the comment was written (parent text). This time we achieved a test accuracy over 50%.

    Finally, as a way to have a third opinion in the voting process suggested by the Condorcet's jury theorem and explored in [7], we used the pretrained model VADER to do sentiment analysis with context. The ideia here was to consider the contrast between the situation and the target sentence, in [6] it was studied the relationship between negative situations/contexts and positive answers/sentences as a way to express sarcasm in the social media Twitter.

## References

[1] Khodak, M., Saunshi, N. and Vodrahalli, K., 2017. A large self-annotated corpus for sarcasm. arXiv preprint arXiv:1704.05579.

[2] Attardo, S. and Raskin, V., 1991. Script theory revis (it) ed: Joke similarity and joke representation model.

[3] Farha, I.A., Oprea, S., Wilson, S. and Magdy, W., 2022, July. Semeval-2022 task 6: isarcasmeval, intended sarcasm detection in english and arabic. In The 16th International Workshop on Semantic Evaluation 2022 (pp. 802-814). Association for Computational Linguistics.

[4] Ashwitha, A., Shruthi, G., Shruthi, H.R., Upadhyaya, M., Ray, A.P. and Manjunath, T.C., 2021. Sarcasm detection in natural language processing. Materials Today: Proceedings, 37, pp.3324-3331.

[5] Adam Paszke et al. “Automatic differentiation in PyTorch”. In: (2017).

[6] Riloff, E., Qadir, A., Surve, P., De Silva, L., Gilbert, N. and Huang, R., 2013, October. Sarcasm as contrast between a positive sentiment and negative situation. In Proceedings of the 2013 conference on empirical methods in natural language processing (pp. 704-714).

[7] Ariharan, V., Eswaran, S.P., Vempati, S. and Anjum, N., 2019. Machine learning quorum decider (MLQD) for large scale IoT deployments. Procedia Computer Science, 151, pp.959-964.